### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by primsi**

### CODE REVIEW

File: dropzonejs.libraries.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+dropzonejs:
```
Comment (primsi) [inline dropzonejs.libraries.yml]: I am not sure about how to deal with the vendor stuff as there is no good way to use composer with js stuff. For now I've put it in /assets/dropzonejs. I was also thinking /vendor/dropzonejs

File: src/Controller/UploadController.php [hunk 2]
```diff
@@ -0,0 +1,174 @@
+<?php
```
Comment (primsi) [inline src/Controller/UploadController.php:1]: A lot of stuff from plupload here - :copyright:  :cat:

File: dropzonejs.info.yml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+name: dropzonejs
+type: module
+description: DropzoneJS
+core: 8.x
+package: Other
```
Comment (slashrsm) [inline dropzonejs.info.yml]: Media?

File: dropzonejs.module [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ * Default template: dropzonejs.html.twig.
+ *
+ * @param array $variables
+ *   An associative array containing:
+ *   - element: A render element representing the file.
+ */
+function template_preprocess_dropzonejs(&$variables) {
+  $element = $variables['element'];
+
+  $variables['attributes'] = [];
```
Comment (slashrsm) [inline dropzonejs.module:49]: What about other attributes?

File: js/dropzone.integration.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  Drupal.behaviors.dropzonejsIntegraion = {
+    attach: function(context) {
+      Dropzone.autoDiscover = false;
+      var selector = $(".dropzone-enable");
+      selector.addClass("dropzone");
+
+      // Initiate dropzonejs.
+      var dropzoneInstance = new Dropzone("#" + selector.attr("id"), {
+        // @todo we should get all this from somewhere.
+        url: "/drupal-8-media/dropzonejs/upload",
```
Comment (slashrsm) [inline js/dropzone.integration.js]: consider base url

File: js/dropzone.integration.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+  Drupal.behaviors.dropzonejsIntegraion = {
+    attach: function(context) {
+      Dropzone.autoDiscover = false;
+      var selector = $(".dropzone-enable");
+      selector.addClass("dropzone");
+
+      // Initiate dropzonejs.
+      var dropzoneInstance = new Dropzone("#" + selector.attr("id"), {
+        // @todo we should get all this from somewhere.
```
Comment (slashrsm) [inline js/dropzone.integration.js]: See how plupload in D7 passes settings to JS

File: js/dropzone.integration.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        // @todo handle files with validaiton errors.
+        uploadedFilesElement.attr('value', currentValue + file.name + ';');
+      });
+
+      // React on file removing.
+      dropzoneInstance.on("removedfile", function(file) {
+        var uploadedFilesElement = selector.siblings(':hidden');
+        var currentValue = uploadedFilesElement.attr('value');
+
+        // Remove the file from the element.
```
Comment (slashrsm) [inline js/dropzone.integration.js:39]: Would be great if we'd delete this file from server too.

File: src/Controller/UploadController.php [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+  /**
+   * Constructs dropzone upload controller route controller.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   Request object.
+   */
+  public function __construct(Request $request) {
+    $this->request = $request;
+    $this->temporaryUploadLocation = \Drupal::config('system.file')->get('path.temporary');
```
Comment (slashrsm) [inline src/Controller/UploadController.php]: inject config?

File: src/Controller/UploadController.php [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+   * Handles DropzoneJs uploads.
+   */
+  public function handleUploads() {
+    // @todo: Implement file_validate_size();
+    try {
+      $this->prepareTemporaryUploadDestination();
+      $this->handleUpload();
+    }
+    catch (UploadException $e) {
+      return $e->getErrorResponse();
```
Comment (slashrsm) [inline src/Controller/UploadController.php:86]: return 500?

File: src/Controller/UploadController.php [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    // Return JSON-RPC response.
+    // @todo Not sure if we need this.
+    return new JsonResponse(
```
Comment (slashrsm) [inline src/Controller/UploadController.php]: Is this from Plupload? Then we probably don't need it or must be adapted to dropzone.

File: src/Controller/UploadController.php [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+   * Reads, checks and return filename of a file being uploaded.
+   *
+   * @param \Symfony\Component\HttpFoundation\File\UploadedFile $file
+   *   An instance of UploadedFile.
+   *
+   * @throws \Drupal\plupload\UploadException
+   */
+  protected function getFilename(UploadedFile $file) {
+    if (empty($this->filename)) {
+      // @todo I am not sure why plupload gets 'name' here ($this->request->request->get('name')).
```
Comment (slashrsm) [inline src/Controller/UploadController.php]: Plupload passes filename as get argument.

File: src/Controller/UploadController.php [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+    if (!$file instanceof UploadedFile) {
+      throw new UploadException(UploadException::REQUEST_ERROR);
+    }
+    elseif ($file->getError() != UPLOAD_ERR_OK) {
+      throw new UploadException(UploadException::FILE_UPLOAD_ERROR);
+    }
+
+    // Open temp file.
+    $tmp = $this->temporaryUploadLocation . $this->getFilename($file);
+    // @todo Is the 'b' at the end of the mode ok?
```
Comment (slashrsm) [inline src/Controller/UploadController.php]: Sure. Binary mode is used to get exact copy of uploaded files.

File: src/Controller/UploadController.php [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+
+    // Append input stream to temp file.
+    while ($buff = fread($in, 4096)) {
+      fwrite($out, $buff);
+    }
+
+    // Be nice and keep everything nice and clean.
+    fclose($in);
+    fclose($out);
+    if ($is_multipart) {
```
Comment (slashrsm) [inline src/Controller/UploadController.php]: Leftover?

File: src/Element/DropzoneJs.php [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+   * @param array $element
+   *   An associative array containing the properties of the element.
+   *   Properties used: #title, #name, #size, #description, #required,
+   *   #attributes.
+   *
+   * @return array
+   *   The $element with prepared variables ready for input.html.twig.
+   */
+  public static function preRenderDropzoneJs($element) {
+    static::setAttributes($element, ['dropzone-enable']);
```
Comment (slashrsm) [inline src/Element/DropzoneJs.php:88]: Should this be class?

File: src/Controller/UploadController.php (hunk 9 above)
Comment (slashrsm) [inline src/Controller/UploadController.php:86]: Meh.... forget.

File: dropzonejs.module (hunk 4 above)
Comment (primsi) [inline dropzonejs.module:49]: Aren't they in $variables['element'] at this point?

File: js/dropzone.integration.js (hunk 7 above)
Comment (primsi) [inline js/dropzone.integration.js:39]: I suggest opening a separate issue for that.

File: src/Element/DropzoneJs.php (hunk 14 above)
Comment (primsi) [inline src/Element/DropzoneJs.php:88]: This sets the class yes.

**Pull request merged**